### PR TITLE
DMA (alternative to #209)

### PR DESCRIPTION
**Pull request by 9names**
Rebased #209 onto current trunk.
Updated to embedded-dma 0.2

Current plan is to drop everything except for SingleChannel, since I think that should end up being simpler and more flexible than the current approach.
To get bidirectional functionality from SingleChannel, will need to update peripherals to provide TX/RX endpoints so that the user can have 2 channels targeting the same peripheral for reading/writing.
For double-buffering I was thinking to pass the second DMA channel as an Option to the constructor, so the same interface is used regardless of single-shot/double buffering.

Will probably rename SingleChannel to Channel or Transfer or something to match the above changes.

EDIT:
  After splitting it up, it looks like it already resembles the above somewhat so this might only require minor changes

Other goals:
- support transfer widths other than 32bit for peripherals
- address concerns with current interface raised in #13 and #209

### CODE REVIEW

**Comment by 9names** [comment]
embedded-dma constrains DMA traits to WORD sizes, where WORD is u8/u16/u32/u64/i8/i16/i32/i64.
64bit sizes compile fine but break at runtime. There are now tests that show this is broken, not sure about the best way to proceed to fix that.

**Comment by 9names** [comment]
Removed target tests from this branch since there's no practical way to keep CI happy with them in there. Will keep them in a separate repo for now.

**Comment by 9names** [comment]
Turns out the solution to the CI problem was to exclude the on-target test project from the workspace.
There's still the rest of the problems that come from having a top-level .cargo/config, but those are at least potentially resolvable.

File: rp2040-hal/src/dma/single_channel.rs
Comment (jannic) [inline rp2040-hal/src/dma/single_channel.rs:182]: How does that interact with DREQ grants? Are they still counted while the channel is disabled?

If they are note, the DREQ counter could get out of sync.

File: rp2040-hal/src/pio.rs
Comment (jannic) [inline rp2040-hal/src/pio.rs:75]: 1?

Comment (9names) [reply rp2040-hal/src/pio.rs:75]: Yeah, makes sense

File: rp2040-hal/src/dma/single_channel.rs
Comment (9names) [reply rp2040-hal/src/dma/single_channel.rs:182]: What do you mean by "DREQ grants"? I can't find this term anywhere.

**Comment by 9names** [comment]
I merged in https://github.com/9names/rp-hal/pull/4 and rebased on latest main, apologies for the noise

File: rp2040-hal/src/dma/single_channel.rs
Comment (jannic) [reply rp2040-hal/src/dma/single_channel.rs:182]: I did some experiments, temporarily pausing a channel and reading DBG_CTDREQ. Looks like the DREQ mechanism keeps counting while a the `en` flag is cleared, so this shoud be fine!

Comment (9names) [reply rp2040-hal/src/dma/single_channel.rs:182]: Ah, good. I read that section of the datasheet + the pico-sdk and it seemed like it would just keep counting, but I was going to test it out to be sure. Now that you've tested it I'll trust that it works. :+1:

**Comment by 9names** [comment]
> Any issues preventing it from being merged?

Well there's at least one issue: it's not finished yet.
The examples are not finished, tests only cover memory-to-memory operations, not all peripherals are implemented yet, and parts of the DMA driver are not complete.
Are you offering to help?

**Comment by jounathaen** [comment]
I could help a little. So, what is left to do?
In the code, I find the following:
- [ ] SPI example TODOS?
- [ ] Ring transfer
- [ ] Drop for `SingleBuffering`
- [ ] Documentation improvements
- [ ] Fix/enhance `pio_dma.rs` example
- [ ] Fix/enhance `spi_dma.rs` example
- [ ] Callbacks?

> The examples are not finished, tests only cover memory-to-memory operations, not all peripherals are implemented yet, 

In any case, we should not wait until everything is polished and final before merging, as a partially working DMA support is still better than no DMA support.

**Comment by 9names** [comment]
MVP as far as I'm concerned:
- Remove support for everything that isn't finished.
- Fix up documentation for everything that is left.
- Add an RTIC example to ensure that it is compatible with resources.
- Add an example that uses a peripheral as a source (ADC would be a good candidate).
- Add an example that does TX and RX on the same peripheral (SPI/UART would be good candidates for this)
- Add an example that does peripheral to peripheral transfers (ADC -> UART would be my choice)
- Add an example that does mem-mem transfers (this can be copy-paste from the on-target tests)
- Verify after we've done all this that everything still works for anyone currently using code from this branch (slint is the only project I'm aware of).

I'm planning on doing a merge from trunk today, then remove the incomplete stuff and start on the docs, if you want to help with the examples that would be much appreciated.

**Comment by 9names** [comment]
To implement Bidirectional support correctly we really need something to start multiple channels simultaneously.
single_channel only allows setting one bit at a time.
https://github.com/9names/rp-hal/blob/7c832ea7f21784c062bd2171ef9a5bd62d6d4776/rp2040-hal/src/dma/single_channel.rs#L206

The examples + tests I've added "work" but are unsafe and inherently racy.

File: rp2040-hal/src/dma/bidirectional.rs
Comment (jounathaen) [inline rp2040-hal/src/dma/bidirectional.rs:1]: Honestly, why do we need the bidirectional config at all?
In your SPI example, you showed that the same can be achieved with two unidirectional configs as well.
This sounds to me like application level code, not HAL level code.

File: rp2040-hal/examples/spi_dma.rs
Comment (jounathaen) [inline rp2040-hal/examples/spi_dma.rs:87]: I'd throw out the bidirectional config in the example anyway. IMHO, the example should be a starting point with "simple" examples, and should not demonstrate all possible uses. If an example for BidirectionalConfig is desired, why not put it in the documentation of that module?

File: rp2040-hal/src/spi.rs
Comment (jounathaen) [inline rp2040-hal/src/spi.rs:56]: One could use https://docs.rs/rp2040-pac/latest/rp2040_pac/dma/ch/ch_ctrl_trig/enum.TREQ_SEL_A.html here, instead of "magic numbers"

File: rp2040-hal/src/dma/mod.rs
Comment (jounathaen) [inline rp2040-hal/src/dma/mod.rs:346]: As this becomes the public API, I'd strongly suggest to use the PAC enum here as well:
https://docs.rs/rp2040-pac/latest/rp2040_pac/dma/ch/ch_ctrl_trig/enum.TREQ_SEL_A.html

Comment (9names) [reply rp2040-hal/src/dma/mod.rs:346]: Yes, having these here pre-dates them being in the PAC.
Should definitely move to using the PAC constants.

File: rp2040-hal/src/dma/bidirectional.rs
Comment (9names) [reply rp2040-hal/src/dma/bidirectional.rs:1]: > This sounds to me like application level code, not HAL level code.

I disagree. Having to resort to fabricating a SPI singleton and then throwing it away after the transaction, and having to do hacks to get both channels started simultaneously is not what something we want to push on our users.
This is enough to test my tests, but I would not recommend it as a general solution to the problem. We can do better.

**Comment by jounathaen** [comment]
Another suggestion: Could you rename `SingleBuffering` to `SingleBufferedTransfer` (and `Bidirectional` accordingly)?
Found the other names a little clumsy when trying it out.

File: rp2040-hal/src/dma/mod.rs
Comment (9names) [reply rp2040-hal/src/dma/mod.rs]: These are now removed

File: rp2040-hal/src/pio.rs
Comment (9names) [reply rp2040-hal/src/pio.rs:78]: Resolved in [573e62e](https://github.com/rp-rs/rp-hal/pull/457/commits/573e62ea42c01be568faee4867e7548f4f8ca2ac)

File: rp2040-hal/src/spi.rs
Comment (9names) [reply rp2040-hal/src/spi.rs:57]: Resolved in [243af46](https://github.com/rp-rs/rp-hal/pull/457/commits/243af465d2e2715ca2a3167b7943e8a2a83d85d2)

**Comment by 9names** [comment]
Bidir is now working for SPI, at least with my trivial tests/examples.
One blocker down.

File: rp2040-hal/src/dma/mod.rs
Comment (jannic) [inline rp2040-hal/src/dma/mod.rs:166]: `StaticReadBuffer` doesn't exist in the latest version of embedded_dma. Should probably be changed to `ReadBuffer::read_buffer`? At least that's how I understand the [changelog](https://github.com/rust-embedded/embedded-dma/blob/master/CHANGELOG.md): "Replace less strict ReadBuffer and WriteBuffer definitions with those of StaticReadBuffer and StaticWriteBuffer. This removes the separate static traits."

File: rp2040-hal/src/dma/mod.rs
Comment (jannic) [inline rp2040-hal/src/dma/mod.rs:187]: I don't understand this safety comment: As `rx_address_count` is not unsafe, what keeps some random safe code from calling this method multiple times?

**Review by jannic — APPROVED** [review]
I still haven't had time to try this code. But as it reads nicely, contains test cases, and has been commented by multiple persons, I'd say we shouldn't wait for a more in-depth review before merging it. It has been waiting long enough.

File: rp2040-hal/src/dma/mod.rs
Comment (9names) [reply rp2040-hal/src/dma/mod.rs:166]: That makes sense with my understanding of the 0.2 changes.
There's a bunch more documentation to fix up as well, unfortunately. Hoping that we'll get more interest for that with the code in-tree than we have with PRs.

File: rp2040-hal/src/dma/mod.rs
Comment (9names) [reply rp2040-hal/src/dma/mod.rs:187]: Comment was from the original PR, but agree that it doesn't make sense.
Maybe the original wasn't going to expose this to end-users?
Also not sure what the side-effect is that was hoped to be avoided, I think I'll write some tests (after merge) to exercise this condition to see if it's a problem.

Comment (9names) [reply rp2040-hal/src/dma/mod.rs:187]: ah.
```rust
/// # Safety
///
/// Once this method has been called, it is unsafe to call any `&mut self`
/// methods on this object as long as the returned value is in use (by DMA).
unsafe fn read_buffer(&self) -> (*const Self::Word, usize);
```
but this is also true if the addresses returned are copied and used for another DMA transfer - the function in and of itself is safe. I'll remove the comment as it's more confusing than helpful.

**Comment by jounathaen** [comment]
I suggest resolving https://github.com/rp-rs/rp-hal/issues/528 before merging, as this would be a major API change afterwards.

**Comment by 9names** [comment]
Alright, it's now updated to the API discussed in #528.
Looks okay to me :) but would appreciate another pair of eyes to review how the API looks @jounathaen

**Comment by jounathaen** [comment]
I think it is very nice now! Couldn't find anything more to complain about :wink: 

Thank you so much for your efforts!

**Comment by 9names** [comment]
I did a test rebase and it looked sensible, so I'm going to trust that Github's "Squash and merge" is going to do the right thing :crossed_fingers:

**Pull request merged**
